cpp/src/arrow/compute/kernels/scalar_temporal_unary.cc

```diff
@@ -1143,6 +1145,117 @@ struct Strftime {
 };
 #endif
 
+// ----------------------------------------------------------------------
+// Convert string representations of timestamps in arbitrary format to timestamps
+
+static std::string GetZone(std::string format) {
+  // Check for use of %z or %Z
+  size_t cur = 0;
+  std::string zone = "";
+  while (cur < format.size() - 1) {
+    if (format[cur] == '%') {
+      if (format[cur + 1] == 'z') {
+        zone = "UTC";
+        break;
+      }
+      cur++;
+    }
+    cur++;
+  }
+  return zone;
+}
+
+template <typename Duration, typename InType>
+struct Strptime {
+  const std::shared_ptr<TimestampParser> parser;
+  const TimeUnit::type unit;
+  const std::string zone;
+  const bool raise_errors;
+
+  static Result<Strptime> Make(KernelContext* ctx, const DataType& type) {
+    const StrptimeOptions& options = StrptimeState::Get(ctx);
+
+    return Strptime{TimestampParser::MakeStrptime(options.format),
+                    std::move(options.unit), GetZone(options.format),
+                    options.raise_errors};
+  }
+
+  static Status Call(KernelContext* ctx, const Scalar& in, Scalar* out) {
+    ARROW_ASSIGN_OR_RAISE(auto self, Make(ctx, *in.type));
+
+    if (in.is_valid) {
+      auto s = internal::UnboxScalar<InType>::Unbox(in);
+      int64_t result;
+      if ((*self.parser)(s.data(), s.size(), self.unit, &result)) {
+        *checked_cast<TimestampScalar*>(out) =
+            TimestampScalar(result, timestamp(self.unit, self.zone));
+      } else {
+        if (self.raise_errors) {
+          return Status::Invalid("Failed to parse string: '", s.data(),
+                                 "' as a scalar of type ",
+                                 TimestampType(self.unit).ToString());
+        } else {
+          out->is_valid = false;
+        }
+      }
+    } else {
+      out->is_valid = false;
+    }
+    return Status::OK();
+  }
+
+  static Status Call(KernelContext* ctx, const ArrayData& in, ArrayData* out) {
+    ARROW_ASSIGN_OR_RAISE(auto self, Make(ctx, *in.type));
+
+    std::unique_ptr<ArrayBuilder> array_builder;
```

Comment:
Using a builder is probably the reason for why we can't `can_write_into_slices`? Since the builder allocates its own buffer. Presumably the kernel has preallocation enabled so you should instead be able to write directly into `out->array()->GetValues<int64_t>(1)`.

Reply:
Done. I used `FirstTimeBitmapWriter` for the null bitmap writing. Please check if it all makes sense.

Reply:
Looks good, though: if the kernel is set to null_handling = INTERSECTION, then instead of setting/clearing every individual bit using the writer, we can just manually clear the bits for slots that fail to parse (and if we are raising errors, we shouldn't need to touch the null bitmap at all) - that should presumably be faster in the common case.

Reply:
Oh, indeed. I've removed the bitmap editing from the `error_is_null == false` case and reduced it in the other case for failed parsing attempts.
Are you saying BitWriter has enough overhead to warrant doing this manually? https://github.com/apache/arrow/blob/master/cpp/src/arrow/util/bitmap_writer.h#L34